### DISCUSSION TITLE
Use external authentication mechanism

### DISCUSSION THREAD

**Discussion by andyrue** [janko/rodauth-rails#69] [Q&A]
My goal is two have two login sections, one for basic users and one for admins. The user login authenticates with an internal db of users stored in the `accounts` table. The admin users would authenticate with an external authentication server. 

I have created an additional configuration for admin accounts, but I can't seem to get everything working quite right. Here is what I've done so far using Jeremy's example of how to authenticate with LDAP found [here](https://github.com/jeremyevans/rodauth#using-ldap-authentication-).

Under the `:admin` config, I added the following things.

```
session_key :login
account_session_value{account}
```

```
account_from_login do |login|
  login
end
```
This returns the username string so I can authenticate with my own mechanism inside the `password_match?` method.

```
password_match? do |password|
  status = ExternalMechanism.authenticate(account, password)
 # This is optional
 # if status['success']
 #   login = account
 #   account = Account.find_or_create_by(email: login)
 #   account.status = 'verified'
 #   account.type = 'Admin'
 #   account.save
 # end
  status['success']
end
```
If the authentication succeeds, then I am (optionally) finding or creating an account for them and making them an Admin.

This all works fine, but where I'm having trouble is with the `current_user(:admin)` method. It tries to lookup the account with a `nil` id, and obviously not finding it, then logging me out. Somehow I need to login the account that gets found/created inside the `password_match?` method. 

I don't really need to have the accounts stored in the database, but I started down that rabbit hole because I couldn't get the `current_user(:admin)` to not want to lookup users in the Account model by their id without, but it doesn't seem to be helping anyway. I also can't authenticate routes that are restricted to the `(:admin)` config. I think this is due to the same problem.

My routes look like this.

```
constraints Rodauth::Rails.authenticated(:admin) do
  namespace :admin do
    resources :resource
  end
end
```

And my initial login check on first page load is:

```
if rodauth(:admin).logged_in?
  render json: { loggedIn: true, account: current_account(:admin) }
end
```

But this results in being logged out due to the response of `current_account(:admin)`.

Thanks for any help!

I feel like I'm almost there, but not sure what I'm doing wrong now.

**Comment by janko** [comment]
If you would prefer not to store authenticated accounts in the database, then you shouldn't be using the `#current_account` method, because that method's only purpose is to retrieve the account from the database. Is there a reason why you want to be using it?

**Comment by andyrue** [comment]
Hah, ya...this appears to be a case of me creating new problems to solve existing ones. I didn't originally have the `#current_account` method there, but added it after trying to create accounts instead. I have since realized the logging out was being caused by my frontend calling `/logout` due to 401 errors being returned. The problem seems to stem from the routing constraints. When my application loads, I check if the user is logged in with `rodauth(:admin).logged_in?` inside `rodauth_controller.rb` which returns the email address of my user as expected. Then I try and access a protected route and at that point `rodauth(:admin).logged_in?` returns `nil` which is unexpected. I know my session isn't being cleared because if I refresh the app, `rodauth(:admin).logged_in?` still returns the user inside `rodauth_controller.rb`. I've tried a few methods of route constraints. Right now I have this in my `rodauth_app.rb` file. But was doing it at the router level before with same result.

```
r.on "admin" do
  r.rodauth(:admin)
  unless rodauth(:admin).logged_in?
    rodauth(:admin).require_authentication
  end

  break # allow the Rails app to handle other "/admin/*" requests
end
```

If you're not seeing anything super obviously wrong, I'll try and put together a sample app so you can take a better look. Thanks!

**Reply by janko** [reply]
Hmm, yeah, I'm not seeing anything wrong here.

**Comment by andyrue** [comment]
Well...this was a very very silly mistake. I was not including `withCredentials` on the frontend request that was failing. * *sigh* * Thank you for your patience. :-)
